Client/game_sa/CDynamicPool.h

```diff
@@ -0,0 +1,171 @@
+/*****************************************************************************
+ *
+ *  PROJECT:     Multi Theft Auto v1.0
+ *  LICENSE:     See LICENSE in the top level directory
+ *  FILE:        game_sa/CDynamicPool.h
+ *  PURPOSE:     Custom implementation for SA pools
+ *
+ *  Multi Theft Auto is available from http://www.multitheftauto.com/
+ *
+ *****************************************************************************/
+
+#pragma once
+
+#include <vector>
+#include <array>
+#include <memory>
+
+template <typename PoolObjT>
+class CDynamicPoolPart
+{
+public:
+    explicit CDynamicPoolPart(std::size_t size) : m_size{size}
+    {
+        m_items = std::make_unique<PoolObjT[]>(size);
+        m_unusedIndexes.reserve(size);
+        for (std::size_t i = 0; i < size; i++)
+            m_unusedIndexes.push_back(i);
+    }
+
+    PoolObjT* AllocateItem()
+    {
+        std::size_t index = m_unusedIndexes.back();
+        m_unusedIndexes.pop_back();
+        return &m_items[index];
+    }
+
+    void RemoveItem(PoolObjT* item)
+    {
+        auto pos = item - m_items.get();
+        m_unusedIndexes.push_back(pos);
+    }
+
+    bool        OwnsItem(PoolObjT* item) const noexcept { return item >= m_items.get() && item < m_items.get() + m_size; }
+    bool        HasFreeSize() const noexcept { return m_unusedIndexes.size() != 0; }
+    std::size_t GetCapacity() const noexcept { return m_size; }
+    std::size_t GetUsedSize() const noexcept { return m_size - m_unusedIndexes.size(); }
+
+private:
+    std::unique_ptr<PoolObjT[]> m_items;
+    std::vector<std::size_t>    m_unusedIndexes;
+    const std::size_t           m_size;
+};
+
+template <std::size_t InitialSize>
+struct PoolGrownByHalfStrategy
+{
+    static constexpr std::size_t GetInitialSize() { return InitialSize; }
+    static constexpr std::size_t GetNextSize(std::size_t index) { return InitialSize / 2; }
+};
+
+template <typename PoolObjT, typename GrownStrategy>
+class CDynamicPool
+{
+public:
+    CDynamicPool()
+    {
+        constexpr size_t initialSize = GrownStrategy::GetInitialSize();
+        m_poolParts.emplace_back(std::make_unique<CDynamicPoolPart<PoolObjT>>(initialSize));
+    }
+
+    PoolObjT* AllocateItem()
+    {
+        for (auto& pool : m_poolParts)
+        {
+            if (pool->HasFreeSize())
+                return pool->AllocateItem();
+        }
+
+        return AllocateNewPart()->AllocateItem();
```

Comment:
An exception std::bad_alloc can be thrown here.

Reply:
Do you have an idea how to handle it?

Reply:
Ideally there should be a way to prevent the game from crashing. I know that this implies many changes so it's out of the scope of this PR. But at least we can put an assertation with a message. Something like

```
try
{
    return AllocateNewPart()->AllocateItem();
}
catch (const std::bad_alloc&)
{
    assert(false && "Could not allocate a memory for single link nodes");
}
```

I believe that this is better than leaving an exception unhandled.